src/librustc_trans/debuginfo/source_loc.rs

```diff
@@ -195,37 +190,42 @@ impl InternalDebugLocation {
     }
 }
 
-pub fn set_debug_location(cx: &CrateContext, debug_location: InternalDebugLocation) {
-    if debug_location == debug_context(cx).current_debug_location.get() {
-        return;
+pub fn set_debug_location(cx: &CrateContext,
+                          builder: Option<llvm::BuilderRef>,
+                          debug_location: InternalDebugLocation) {
+    if builder.is_none() {
+        if debug_location == debug_context(cx).current_debug_location.get() {
+            return;
+        }
     }
 
-    let metadata_node;
-
-    match debug_location {
+    let metadata_node = match debug_location {
         KnownLocation { scope, line, .. } => {
             // Always set the column to zero like Clang and GCC
             let col = UNKNOWN_COLUMN_NUMBER;
             debug!("setting debug location to {} {}", line, col);
 
             unsafe {
-                metadata_node = llvm::LLVMDIBuilderCreateDebugLocation(
+                llvm::LLVMDIBuilderCreateDebugLocation(
                     debug_context(cx).llcontext,
                     line as c_uint,
                     col as c_uint,
                     scope,
-                    ptr::null_mut());
+                    ptr::null_mut())
             }
         }
         UnknownLocation => {
             debug!("clearing debug location ");
-            metadata_node = ptr::null_mut();
+            ptr::null_mut()
         }
     };
 
-    unsafe {
-        llvm::LLVMSetCurrentDebugLocation(cx.raw_builder(), metadata_node);
+    if builder.is_none() {
```

Comment:
Why is it necessary to make this conditional? It seems to me that if this condition is removed, then in MIR we can avoid doing both `apply_to_bcx` and `apply` in certain cases (only `apply_to_bcx` would be enough)